Sources/Subscription/Managers/SubscriptionManager.swift

```diff
@@ -171,20 +282,166 @@ public final class DefaultSubscriptionManager: SubscriptionManager {
+        try await subscriptionEndpointService.ingestSubscription(confirmation.subscription)
+        Logger.subscription.log("Purchase confirmed!")
+        return confirmation.subscription
+    }
+
+    // MARK: - Features
+
+    /// Returns the features available for the current subscription, a feature is enabled only if the user has the corresponding entitlement
+    /// - Parameter forceRefresh: ignore subscription and token cache and re-download everything
+    /// - Returns: An Array of SubscriptionFeature where each feature is enabled or disabled based on the user entitlements
+    public func currentSubscriptionFeatures(forceRefresh: Bool) async -> [SubscriptionFeature] {
+        guard isUserAuthenticated else { return [] }
+
+        if let subscriptionFeatureFlagger,
+           subscriptionFeatureFlagger.isFeatureOn(.isLaunchedROW) || subscriptionFeatureFlagger.isFeatureOn(.isLaunchedROWOverride) {
+            do {
+                let currentSubscription = try await getSubscription(cachePolicy: .returnCacheDataDontLoad)
+                let tokenContainer = try await getTokenContainer(policy: forceRefresh ? .localForceRefresh : .local)
+                let userEntitlements = tokenContainer.decodedAccessToken.subscriptionEntitlements
+                let availableFeatures = currentSubscription.features ?? [] // await subscriptionFeatureMappingCache.subscriptionFeatures(for: subscription.productId)
+
+                // Filter out the features that are not available because the user doesn't have the right entitlements
+                let result = availableFeatures.map({ featureEntitlement in
+                    let enabled = userEntitlements.contains(featureEntitlement)
+                    return SubscriptionFeature(entitlement: featureEntitlement, enabled: enabled)
+                })
+                Logger.subscription.log("""
+User entitlements: \(userEntitlements)
+Available Features: \(availableFeatures)
+Subscription features: \(result)
+""")
+                return result
+            } catch {
                 return []
             }
         } else {
-            return [.networkProtection, .dataBrokerProtection, .identityTheftRestoration]
+            let result = [SubscriptionFeature(entitlement: .networkProtection, enabled: true),
+                          SubscriptionFeature(entitlement: .dataBrokerProtection, enabled: true),
+                          SubscriptionFeature(entitlement: .identityTheftRestoration, enabled: true)]
+            Logger.subscription.debug("Default Subscription features: \(result)")
+            return result
+        }
+    }
+
+    public func isFeatureActive(_ entitlement: SubscriptionEntitlement) async -> Bool {
+        let currentFeatures = await currentSubscriptionFeatures(forceRefresh: false)
+        return currentFeatures.contains { feature in
+            feature.entitlement == entitlement && feature.enabled
```

Comment:
Similar as comment above. I strongly would encourage to stick to naming conventions already established and rename it to `hasEntitlement(_:)`. Also there should be no logic crosschecking it with the entitlements here, just simple check for entitlements array if one is present and return true or false.
I think I see what you tried to achieve here but the whole quite simple concept becomes convoluted without reason. `Features` in the subscription define what UI-wise is part of current subscription, and `Entitlements` define which parts user currently has access to within the scope of current subscription feature set.

Reply:
As explained above `hasEntitlement(_:)` doesn't make sense anymore, nobody outside SubscriptionManager can or should know/care about internal implementation details (The S of SOLID principles) so it's the duty of `currentSubscriptionFeatures` to provide the right information needed from the UI, we can't ask the users (main apps) to check if the use entitlement is present, get the subscription entitlements, compare them and then update the UI.